File: 2021_12.py
import sys;
import logging;

AOC_DAY=12
AOC_YEAR=2021

class cave():
    def __init__(self, name):
        self.name = name;
        self.big = False;
        self.start = False;
        self.end = False;
        if(self.name[0].upper() == self.name[0]):
            self.big = True;
        elif(self.name == 'start'):
            self.start = True;
        elif(self.name == 'end'):
            self.end = True;
        self.connected_to = []

    def add_connected(self, c):
        self.connected_to.append(c)
        self.connected_to.sort();

    def find_next(self, path):
        # return a sorted list of options for the next cave to visit
        retval = [];
        for i in self.connected_to:
            if i.big or not i in path:
                retval.append(i);
        return retval;

    def __eq__(self, other):
        if type(other) == str:
            return self.name == other;
        return self.name == other.name;
    def __lt__(self, other):
        if(self.end or other.start):
            return False;
        if(other.end or self.start):
            return True;
        return self.name < other.name;

    def __repr__(self):
        return self.name;


def count_paths(start, path, logger):
    npaths = 0;
    for next_path in start.find_next(path):
        if(next_path.end):
            npaths += 1;
            logger.debug("found path {}".format(path + [next_path]));
        else:
            npaths += count_paths(next_path, path + [next_path], logger);

    return npaths;

def read_caves():
    caves = [];
    cave_names = [];
    start = None;
    line = sys.stdin.readline().strip();
    while(line):
        (cave1_name, cave2_name) = line.split('-');
        try:
            cave1 = caves[caves.index(cave1_name)]
        except Exception as e:
            cave1 = cave(cave1_name);
            caves.append(cave1);
        try:
            cave2 = caves[caves.index(cave2_name)]
        except:
            cave2 = cave(cave2_name);
            caves.append(cave2);
        cave1.add_connected(cave2);
        cave2.add_connected(cave1);
        if start is None and cave1.start:
            start = cave1;
        elif start is None and cave2.start:
            start = cave2;
        line = sys.stdin.readline().strip();

    print(caves);

    return start;

if('__main__' == __name__):
    import argparse;

    # deal with command line arguments
    parser = argparse.ArgumentParser(
            description = 'Advent of Code {} day {} solution'.format(
                AOC_YEAR, AOC_DAY));
    logger = logging.Logger('AOC{}d{}'.format(AOC_YEAR, AOC_DAY));
    logger.addHandler(logging.StreamHandler());
    logger.setLevel(logging.INFO);

    parser.add_argument(
        '-2', '--part2', action='store_true',
        help='part 2?', default=False);
    parser.add_argument(
        '-v', '--verbose', action='store_true',
        help='be verbose', default=False);

    args = parser.parse_args();
    if(args.verbose):
        logger.setLevel(logging.DEBUG);

    start = read_caves();
    npaths = count_paths(start, [start], logger);
    print(npaths);

File: 2021_12.rb
require 'logger'

AOCDAY = 12
AOCYEAR = 2021

class Cave
  attr_accessor :name
  attr_accessor :small

  def initialize(name, logger)
    @name = name
    @logger = logger
    @small = false
    if name.downcase == name
      @small = true
    end
    @connections = []
  end

  def add_connection(other)
    @connections << other
    @connections.sort()
  end

  def <=>(other)
    @name <=> other.name
  end

  def to_s()
    return @name
  end

  def get_n_paths(path, part2: false)
    small_cave_visits_allowed = part2 ? 2:1
    npaths = 0

    # for each cave we're connected to
    @connections.each do |nextcave|
      @logger.debug("Cave<#{@name}> checking connection #{nextcave}");
      # check to see if this is a valid next step
      if nextcave.name == 'start'
        #   start can only be visited once
      elsif nextcave.name == 'end'
        #   ends this path
        @logger.debug("Path: #{path.map(&:to_s).join(',')}");
        npaths += 1
      elsif nextcave.small
        #   small caves can't be visited more than once
        nextcaves_in_path =  path.select {|x| x.name == nextcave.name }
        if nextcaves_in_path.length < small_cave_visits_allowed
          npaths += nextcave.get_n_paths(path.clone() + [nextcave], part2: part2);
        end
      else
        #   large cave. Do it.
        npaths += nextcave.get_n_paths(path.clone() + [nextcave], part2: part2);
      end
    end
    npaths
  end
end

def run(inputfile, logger, **kwargs)
  caves = {}

  File.open(inputfile).each_line do |line|
    # read two names
    first,last = line.strip.split('-')
    # figure out if we have them already
    if caves.has_key? first
      first_cave = caves[first]
    else
      first_cave = caves[first] = Cave.new(first, logger)
    end
    if caves.has_key? last
      last_cave = caves[last]
    else
      last_cave = caves[last] = Cave.new(last, logger)
    end

    first_cave.add_connection(last_cave)
    last_cave.add_connection(first_cave)
  end

  # beginning at start position, traverse caves
  start = caves["start"]
  npaths = start.get_n_paths([start], part2: kwargs.include?(:part2))
  puts("Found #{npaths} paths!");
end


if __FILE__ == $0
  require 'optparse'

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: $0 [options] <inputfile> [<more inputfiles>]\n" \
      "  Advent of Code #{AOCYEAR} day #{AOCDAY}"

    opts.on("-v", "--verbose", "Be verbose") do |v|
      options[:verbose] = v
    end

    opts.on("-2", "--part2", "Run part 2") do |part2|
      options[:part2] = true
    end

    opts.on("-h", "--help", "Print this help message") do
      puts opts
      exit
    end
  end.parse!

  logger = Logger.new(STDOUT);
  logger.level = (options.include? :verbose) ? Logger::DEBUG : Logger::WARN

  ARGV.each do |inputfile|
    run(inputfile, logger, **options)
  end
end
